File: python/examples/vtk/isocontours.py
"""
Interactive isocontour extraction example with VTK

Features:
- Extract multiple isocontours from scalar field on mesh
- Scroll wheel (with Shift) to move isocontour levels
- Press 'n' to randomize the cutting plane
- Real-time performance metrics for isocontour extraction

Usage:
    python isocontours.py mesh.stl

The scalar field is computed as signed distance from a random plane.
Multiple isocontour levels are extracted simultaneously and visualized as tubes.
Extraction time is averaged over the last 100 frames.
"""
import sys
import os
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../'))

import vtk
from vtk.util import numpy_support
import numpy as np
import trueform as tf

# Import utilities
from util import (
    MeshData,
    numpy_to_polydata,
    compute_centering_and_scaling_transform,
    curves_to_polydata,
    BaseInteractor,
    RollingAverage,
    format_time_ms,
    create_text_actor,
    create_renderer_with_text_strip,
)


class IsocontourInteractor(BaseInteractor):
    """Interactor for isocontour visualization with scrolling and randomization"""

    def __init__(self, mesh_data, curve_polydata, text_actor):
        super().__init__()

        # Keep mesh data (wraps trueform mesh + actor)
        self.mesh_data = mesh_data
        self.mesh = mesh_data.mesh  # Convenience reference
        self.curve_polydata = curve_polydata
        self.text_actor = text_actor

        # Scalar field data
        self.scalars = None
        self.distance = 0.0  # Offset for isocontour levels
        self.min_d = 0.0     # Min scalar value
        self.max_d = 0.0     # Max scalar value

        # Timing
        self.times = RollingAverage(maxlen=100)

        # Initialize random plane (but don't compute curves yet - interactor not attached)
        self.reset_plane()

        # Add custom event handlers for scrolling and keys
        self.AddObserver("KeyPressEvent", self.on_key_press)
        self.AddObserver("MouseWheelForwardEvent", self.on_mouse_wheel_forward)
        self.AddObserver("MouseWheelBackwardEvent", self.on_mouse_wheel_backward)

    def reset_plane(self):
        """Create a new random plane and compute scalar field"""
        # Create random plane through a random point
        random_normal = np.random.randn(3).astype(np.float32)
        random_normal /= np.linalg.norm(random_normal)
        random_point = self.mesh.points[np.random.randint(0, len(self.mesh.points))]

        plane = tf.Plane.from_point_normal(random_point, random_normal)

        # Compute scalar field (signed distance to plane)
        self.scalars = tf.distance_field(self.mesh.points, plane)
        self.distance = 0.0

        # Cache min/max for adaptive level spacing
        self.min_d = float(np.min(self.scalars))
        self.max_d = float(np.max(self.scalars))

    def compute_curves(self):
        """Extract isocontours and update visualization"""
        import time as time_module

        # Compute adaptive threshold levels (same algorithm as isobands.cpp)
        N = 10  # Number of levels

        # Compute spacing based on scalar field range
        s = (self.max_d - self.min_d) / float(N)

        # Find which segment k the current distance falls into
        a = (self.distance - self.min_d) / s
        k = int(np.floor(a))
        k = np.clip(k, 0, N - 1)

        # Create N levels centered around current distance
        # This ensures cutvalues[k] == distance exactly
        thresholds = np.array([self.distance + (i - k) * s for i in range(N)], dtype=np.float32)

        # Extract isocontours with timing
        start_time = time_module.perf_counter()
        paths, curve_points = tf.isocontours(self.mesh, self.scalars, thresholds)
        elapsed = time_module.perf_counter() - start_time

        # Update timing
        self.times.add(elapsed)
        avg_time = self.times.get_average()
        self.text_actor.SetInput(f"Isocontours time: {format_time_ms(avg_time)}")

        # Convert curves to polydata (no tube filter needed!)
        if len(paths) > 0 and len(curve_points) > 0:
            poly = curves_to_polydata(paths, curve_points)
            self.curve_polydata.ShallowCopy(poly)
        else:
            # No curves - clear visualization
            self.curve_polydata.Initialize()

        self.curve_polydata.Modified()

        # Only render if interactor is attached
        if self.GetInteractor() is not None:
            self.GetInteractor().Render()

    def on_key_press(self, obj, event):
        """Handle key press events"""
        key = self.GetInteractor().GetKeySym()
        if key == "n":
            # Randomize plane
            self.reset_plane()
            self.compute_curves()
        else:
            # Pass to base class for camera controls
            vtk.vtkInteractorStyleTrackballCamera.OnKeyPress(self)

    def on_mouse_wheel_forward(self, obj, event):
        """Handle mouse wheel forward"""
        if self.GetInteractor().GetShiftKey():
            # Move isocontours forward
            self.distance += 0.05
            self.compute_curves()
        else:
            # Pass to base class for zoom
            vtk.vtkInteractorStyleTrackballCamera.OnMouseWheelForward(self)

    def on_mouse_wheel_backward(self, obj, event):
        """Handle mouse wheel backward"""
        if self.GetInteractor().GetShiftKey():
            # Move isocontours backward
            self.distance -= 0.05
            self.compute_curves()
        else:
            # Pass to base class for zoom
            vtk.vtkInteractorStyleTrackballCamera.OnMouseWheelBackward(self)


def main():
    # Parse command line arguments
    if len(sys.argv) < 2:
        print("Usage: python isocontours.py mesh.stl")
        sys.exit(1)

    mesh_file = sys.argv[1]

    # Load mesh
    faces, points = tf.read_stl(mesh_file)

    # Center and scale mesh
    transform = compute_centering_and_scaling_transform(points, target_radius=10.0)

    # Apply transformation
    points_homogeneous = np.hstack([points, np.ones((len(points), 1), dtype=points.dtype)])
    points_transformed = (transform @ points_homogeneous.T).T[:, :3]

    # Create trueform mesh FIRST (this is the primary data structure)
    mesh = tf.Mesh(faces, points_transformed)
    mesh.build_tree()  # Build spatial tree for ray casting

    # Create VTK polydata for visualization (just a view)
    poly = numpy_to_polydata(points_transformed, faces)

    # Create renderers (main + text strip)
    renderer, renderer_text = create_renderer_with_text_strip()

    # Setup mesh actor
    mapper = vtk.vtkPolyDataMapper()
    mapper.SetInputData(poly)
    actor = vtk.vtkActor()
    actor.SetMapper(mapper)
    actor.GetProperty().SetColor(0.8, 0.8, 0.8)
    renderer.AddActor(actor)

    # Create VTK matrix for transformations
    matrix = vtk.vtkMatrix4x4()
    matrix.Identity()
    actor.SetUserMatrix(matrix)

    # Wrap in MeshData for interaction
    mesh_data = MeshData(mesh, actor, matrix)

    # Setup curve actor (initially empty)
    curve_poly = vtk.vtkPolyData()
    curve_poly.Initialize()
    curve_mapper = vtk.vtkPolyDataMapper()
    curve_mapper.SetInputData(curve_poly)
    curve_actor = vtk.vtkActor()
    curve_actor.SetMapper(curve_mapper)

    # Render lines as tubes (GPU-accelerated, no geometry generation needed)
    curve_actor.GetProperty().SetColor(1.0, 0.1, 0.1)  # Red curves
    curve_actor.GetProperty().SetRenderLinesAsTubes(True)
    curve_actor.GetProperty().SetLineWidth(8.0)  # Tube width in pixels

    renderer.AddActor(curve_actor)

    # Create text actors for bottom strip
    text_time = create_text_actor(
        "Isocontours time: 0 ms",
        font_size=38,
        position=(0.03, 0.30),
        justification='left'
    )
    renderer_text.AddViewProp(text_time)

    text_instructions = create_text_actor(
        "Press N to randomize plane",
        font_size=38,
        position=(0.03, 0.70),
        justification='left'
    )
    renderer_text.AddViewProp(text_instructions)

    text_help = create_text_actor(
        "Hold Shift and scroll to move\n\nPowered by trueform",
        font_size=38,
        position=(0.97, 0.55),
        justification='right'
    )
    text_help.GetTextProperty().SetLineSpacing(1.4)
    renderer_text.AddViewProp(text_help)

    # Setup render window and interactor
    render_window = vtk.vtkRenderWindow()
    render_window.AddRenderer(renderer)
    render_window.AddRenderer(renderer_text)
    render_window.SetSize(800, 600)

    interactor = vtk.vtkRenderWindowInteractor()
    interactor.SetRenderWindow(render_window)

    # Setup custom interactor style
    style = IsocontourInteractor(mesh_data, curve_poly, text_time)
    interactor.SetInteractorStyle(style)

    # Compute initial curves (now that interactor is attached)
    style.compute_curves()

    # Start
    render_window.Render()
    interactor.Start()


if __name__ == "__main__":
    main()

File: cpp/examples/vtk/isocontours.cpp
#include "util/vtk_examples.hpp"

#include <trueform/cpp/core/build_tree.hpp>
#include <trueform/cpp/core/cache.hpp>
#include <trueform/cpp/core/index_type.hpp>
#include <trueform/cpp/core/mesh.hpp>
#include <trueform/cpp/iso/isocontours.hpp>
#include <trueform/cpp/spatial/distance.hpp>
#include <trueform/cpp/spatial/primitive.hpp>

#include <vtkActor.h>
#include <vtkInteractorStyleTrackballCamera.h>
#include <vtkMatrix4x4.h>
#include <vtkObjectFactory.h>
#include <vtkPolyData.h>
#include <vtkProperty.h>
#include <vtkRenderWindow.h>
#include <vtkRenderWindowInteractor.h>
#include <vtkRenderer.h>
#include <vtkSmartPointer.h>
#include <vtkTextActor.h>

#include <algorithm>
#include <array>
#include <chrono>
#include <cmath>
#include <cstddef>
#include <iostream>
#include <limits>
#include <random>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

namespace examples = trueform::vtk_examples;
/// A field is cut on the mesh the reader gave, and a mixed carrier is what a
/// mesh read from a file is.
using index_t = tf::cpp::default_index_t;
using mixed_storage = tf::polygons_buffer<index_t, float, 3, tf::dynamic_size>;
using mixed_cache = tf::cpp::cache<index_t, float, 3, tf::dynamic_size>;
using mixed_mesh = tf::cpp::mesh<index_t, float, 3, tf::dynamic_size>;

/// What this example holds: the storage the reader gave and the cache that
/// remembers it. `mesh()` is the assembly every entry takes.
struct surface_source {
  mixed_storage polygons;
  mutable mixed_cache cache;

  auto mesh() const -> mixed_mesh {
    return {polygons.faces(), polygons.points(), cache};
  }
};

class isocontour_interactor final : public vtkInteractorStyleTrackballCamera {
public:
  static auto New() -> isocontour_interactor *;
  vtkTypeMacro(isocontour_interactor, vtkInteractorStyleTrackballCamera);

  auto initialize(const surface_source &value, vtkPolyData *curves,
                  vtkTextActor *timing) -> void {
    _surface = &value;
    _curve_polydata = curves;
    _timing = timing;
    reset_plane();
  }

  auto compute_curves() -> void {
    constexpr int number_of_levels = 10;
    auto wrapped_offset = std::fmod(_distance, _spacing);
    if (wrapped_offset < 0.0F)
      wrapped_offset += _spacing;

    std::vector<float> values;
    values.reserve(number_of_levels + 3);
    for (int i = -1; i != number_of_levels + 2; ++i)
      values.push_back(_minimum + wrapped_offset +
                       static_cast<float>(i) * _spacing);
    const auto thresholds =
        examples::make_array(std::move(values), {number_of_levels + 3});

    const auto start = std::chrono::steady_clock::now();
    const auto curves =
        tf::cpp::isocontours(_surface->mesh(), _scalars, thresholds);
    _times.add(examples::elapsed_seconds(start));
    const auto label = std::string("Isocontours time: ") +
                       examples::format_milliseconds(_times.average());
    _timing->SetInput(label.c_str());

    const auto polydata = examples::to_vtk_polydata(curves);
    examples::replace_polydata(_curve_polydata, polydata);
    if (this->Interactor)
      this->Interactor->Render();
  }

  auto OnKeyPress() -> void override {
    if (std::string(this->Interactor->GetKeySym()) == "n") {
      reset_plane();
      compute_curves();
      return;
    }
    vtkInteractorStyleTrackballCamera::OnKeyPress();
  }

  auto OnMouseWheelForward() -> void override {
    if (this->Interactor->GetControlKey()) {
      _distance += _spacing * 0.1F;
      compute_curves();
      return;
    }
    vtkInteractorStyleTrackballCamera::OnMouseWheelForward();
  }

  auto OnMouseWheelBackward() -> void override {
    if (this->Interactor->GetControlKey()) {
      _distance -= _spacing * 0.1F;
      compute_curves();
      return;
    }
    vtkInteractorStyleTrackballCamera::OnMouseWheelBackward();
  }

protected:
  isocontour_interactor() = default;
  ~isocontour_interactor() override = default;

private:
  auto reset_plane() -> void {
    if (!_surface || _surface->polygons.points_buffer().size() == 0)
      throw std::invalid_argument("isocontours requires a non-empty mesh");

    std::normal_distribution<float> normal_distribution(0.0F, 1.0F);
    examples::point3<float> normal{normal_distribution(_generator),
                                   normal_distribution(_generator),
                                   normal_distribution(_generator)};
    const auto normal_length = std::sqrt(
        normal[0] * normal[0] + normal[1] * normal[1] + normal[2] * normal[2]);
    for (auto &coordinate : normal)
      coordinate /= normal_length;

    const auto points = _surface->polygons.points();
    std::uniform_int_distribution<std::size_t> point_distribution(
        0, points.size() - 1);
    const auto source = points[point_distribution(_generator)];
    set_plane({source[0], source[1], source[2]}, normal);
  }

  auto set_plane(const examples::point3<float> &point,
                 const examples::point3<float> &normal) -> void {
    const auto plane_d =
        -(normal[0] * point[0] + normal[1] * point[1] + normal[2] * point[2]);
    const auto plane = tf::cpp::primitive<float>(
        tf::cpp::primitive_kind::plane,
        examples::make_array<float>({normal[0], normal[1], normal[2], plane_d},
                                    {4}));
    const auto points = tf::cpp::primitive<float>(
        tf::cpp::primitive_kind::point,
        examples::points_array(_surface->polygons.points_buffer()));
    _scalars = tf::cpp::distance(points, plane).batch();
    _distance = 0.0F;
    const auto limits = std::minmax_element(_scalars.begin(), _scalars.end());
    _minimum = *limits.first;
    _maximum = *limits.second;
    _spacing = (_maximum - _minimum) / 10.0F;
  }

  const surface_source *_surface = nullptr;
  vtkPolyData *_curve_polydata = nullptr;
  vtkTextActor *_timing = nullptr;
  tf::cpp::nd_array<float> _scalars;
  float _distance = 0.0F;
  float _minimum = 0.0F;
  float _maximum = 0.0F;
  float _spacing = 0.1F;
  examples::rolling_average _times{100};
  std::mt19937 _generator{std::random_device{}()};
};

vtkStandardNewMacro(isocontour_interactor);

} // namespace

int main(int argc, char **argv) {
  try {
    const auto help = argc == 2 && (std::string(argv[1]) == "-h" ||
                                    std::string(argv[1]) == "--help");
    if (argc != 2 || help) {
      auto &stream = help ? std::cout : std::cerr;
      stream << "Usage: " << argv[0] << " mesh.stl\n\n"
             << "Controls:\n"
             << "  N              Randomize cutting plane\n"
             << "  Ctrl + Scroll  Move isocontour levels\n"
             << "  Mouse drag     Rotate camera\n";
      return help ? 0 : 2;
    }

    surface_source surface{examples::as_mixed_mesh(tf::cpp::read_stl(argv[1])),
                           {}};
    examples::materialize_center_and_scale(surface.polygons, 10.0F);
    tf::cpp::build_tree(surface.mesh());
    const auto source_polydata = examples::to_vtk_polydata(surface.polygons);

    auto renderer = vtkSmartPointer<vtkRenderer>::New();
    const auto text_renderer = examples::make_text_strip(renderer);

    const auto mesh_actor = examples::make_actor(source_polydata);
    mesh_actor->GetProperty()->SetColor(0.8, 0.8, 0.8);
    auto identity = vtkSmartPointer<vtkMatrix4x4>::New();
    identity->Identity();
    mesh_actor->SetUserMatrix(identity);
    renderer->AddActor(mesh_actor);

    auto curve_polydata = vtkSmartPointer<vtkPolyData>::New();
    curve_polydata->Initialize();
    const auto curve_actor = examples::make_actor(curve_polydata);
    curve_actor->GetProperty()->SetColor(1.0, 0.1, 0.1);
    curve_actor->GetProperty()->SetRenderLinesAsTubes(true);
    curve_actor->GetProperty()->SetLineWidth(8.0);
    renderer->AddActor(curve_actor);

    const auto timing =
        examples::make_text_actor("Isocontours time: 0 ms", 38, {0.03, 0.5});
    text_renderer->AddViewProp(timing);

    auto window = vtkSmartPointer<vtkRenderWindow>::New();
    window->AddRenderer(renderer);
    window->AddRenderer(text_renderer);
    window->SetSize(800, 600);

    auto interactor = vtkSmartPointer<vtkRenderWindowInteractor>::New();
    interactor->SetRenderWindow(window);
    auto style = vtkSmartPointer<isocontour_interactor>::New();
    style->initialize(surface, curve_polydata, timing);
    interactor->SetInteractorStyle(style);

    style->compute_curves();
    window->Render();
    interactor->Start();
    return 0;
  } catch (const std::exception &error) {
    std::cerr << "isocontours: " << error.what() << '\n';
    return 1;
  }
}
